quickwit-indexing/src/source/kafka_source.rs

```diff
@@ -296,86 +304,78 @@ impl Source for KafkaSource {
+                        docs.push(doc);
                     } else {
-                        // TODO: this is probably not what should happen here.
-                        // In this case there is no current data in the topic but on a live stream
-                        // more data may come. This should ideally wait for a period of time for new
-                        // data to arrive before returning.
+                        self.state.num_invalid_messages += 1;
+                    }
+                    batch_num_bytes += message.payload_len() as u64;
+                    self.state.num_bytes_processed += message.payload_len() as u64;
+                    self.state.num_messages_processed += 1;
+
+                    let partition_id = self
+                        .state
+                        .assigned_partition_ids
+                        .get(&message.partition())
+                        .ok_or_else(|| {
+                            anyhow::anyhow!(
+                                "Received message from unassigned partition `{}`. Assigned partitions: \
+                                 `{{{}}}`.",
+                                message.partition(),
+                                self.state.assigned_partition_ids.keys().join(", "),
+                            )
+                        })?
+                        .clone();
+                    let current_position = Position::from(message.offset());
+                    let previous_position = self
+                        .state
+                        .current_positions
+                        .insert(message.partition(), current_position.clone())
+                        .unwrap_or_else(|| previous_position_for_offset(message.offset()));
+                    checkpoint_delta
+                        .record_partition_delta(partition_id, previous_position, current_position)
+                        .context("Failed to record partition delta.")?;
+
+                    if batch_num_bytes >= TARGET_BATCH_NUM_BYTES {
                         break;
                     }
+                    ctx.record_progress();
+                }
+                _ = &mut deadline => {
+                    break;
                 }
             }
         }
-
         if !checkpoint_delta.is_empty() {
             let batch = RawDocBatch {
                 docs,
                 checkpoint_delta,
             };
             ctx.send_message(batch_sink, batch).await?;
         }
-
-        // TODO: not entirely sure why this is here. Seems like this would prevent stream processing
-        if self.state.num_active_partitions == 0 {
-            info!(topic = %self.topic, "Reached end of topic.");
-            ctx.send_exit_with_success(batch_sink).await?;
-            return Err(ActorExitStatus::Success);
-        }
-
+        // TODO: Find a better way to stop the source for testing, this not longer works with consumer groups
```

Comment:
We can always stop an actor and performs some assert, but that means that we need to first sleep long enough for the actor to do his work before sending the stop command. The beauty of stopping when partitions are no longer active is that we can just wait for the actor to exit gracefully and the test takes the minimum amount of time required.

Reply:
I was looking at adding a `streaming` bool parameter to `KafkaSourceParams`. This would default to `true` and then if set to `false` would set `enable.partition.eof:true` and then enable this block of code to exit once all partitions are caught up. Would also require a check to wait for the assignments from the initial rebalance before becoming active and then the tests should be reliable (I think). 

This would also provide the ability to run workloads to simply load a topic and then exit the source without continuing to poll it. We do this fairly often in various scenarios.

Reply:
Agreed. Something similar already exists for the Kinesis source. I'll push something tomorrow.